# Methanococcus maripaludis S2 and Desulfovibrio vulgaris Hildenborough co-cultures
In this project, we used syntrophic cocultures of the sulfate-reducing bacterium *Desulfovibrio vulgaris Hildenborough* and the methanogenic archaeon *Methanococcus maripaludis S2*. In addition to the monocultures of both species (Dvu & Mmp) a non-evolved ancestor coculture (Anc-Co, coculture right after mixing species)and two evolved cocultures after 300 & 1000 generations (300-Co & 1000-Co) were analyzed by transcriptomics. These data were used to identify differentially expressed genes connected to cell-to-cell interactions, as well as membrane lipid biosynthesis to determine changes in the cell membrane of the syntrophic coculture partners. To investigate the impact of the evolutionary adaptation and obtained mutations, both partners were re-isolated from all three cocultures. Occurred mutations based on adaptation to syntrophic interactions were identified by genome sequencing and variants analysis.

In this repository, you can find the scripts used for the analyses and some of the supplementary or intermediate files.
